Instance crashed while SELECT including both AVG and GROUP BY

CREATE TABLE t1 (
  spID int unsigned,
  userID int unsigned,
  score smallint unsigned
);

INSERT INTO t1 VALUES (1,1,1);
INSERT INTO t1 VALUES (2,2,2);
INSERT INTO t1 VALUES (2,1,4);
INSERT INTO t1 VALUES (3,3,3);
INSERT INTO t1 VALUES (1,1,5);

mysql> select * from t1;
+------+--------+-------+
| spID | userID | score |
+------+--------+-------+
|    3 |      3 |     3 |
|    2 |      2 |     2 |
|    1 |      1 |     5 |
|    1 |      1 |     1 |
|    2 |      1 |     4 |
+------+--------+-------+
5 rows in set (0.00 sec)


mysql> SELECT userID, AVG(score) FROM t1 WHERE spID=2  GROUP BY userID;
ERROR 2013 (HY000): Lost connection to MySQL server during query


-------------------------------------------------------------------------------------
query:SELECT userID, AVG(score) FROM t1 WHERE spID=2  GROUP BY userID 
o: t1 -> π([spID -> spID, userID -> userID, score -> score] -> σ(spID = 2) -> γ([userID], [avg(score) -> avg(score)]) -> π([userID -> userID, avg(score) -> avg(score)]
panic: interface conversion: interface {} is [][]interface {}, not []interface {}

goroutine 13 [running]:
matrixone/pkg/sql/colexec/group.(*Container).eval(0xc0000b05b8, 0x1, 0xc0003b0840, 0x1, 0x1, 0xc0003ee360, 0x3, 0xc0003b0840, 0x1, 0x1, ...)
	/root/matrixone-main/pkg/sql/colexec/group/group.go:436 +0x44c5
matrixone/pkg/sql/colexec/group.Call(0xc0003ee360, 0xa81d40, 0xc0000b05a0, 0xc00019cc00, 0x0, 0x0)
	/root/matrixone-main/pkg/sql/colexec/group/group.go:166 +0xdba
matrixone/pkg/vm.Run(0xc0003f07e0, 0x4, 0x4, 0xc0003ee360, 0xc00001e6f0, 0x3, 0x3)
	/root/matrixone-main/pkg/vm/vm.go:35 +0x9b
matrixone/pkg/vm/pipeline.(*Pipeline).Run(0xc0001a9e70, 0xc00001a280, 0x1, 0x1, 0xc0003ee360, 0x0, 0x0, 0x0)
	/root/matrixone-main/pkg/vm/pipeline/pipeline.go:62 +0x26f
matrixone/pkg/sql/compile.(*Scope).Run(0xc0003f04e0, 0xc0e3a0, 0xc00032b620, 0x0, 0x0)
	/root/matrixone-main/pkg/sql/compile/compile.go:261 +0x6af
matrixone/pkg/sql/compile.(*Scope).MergeRun.func1(0xc0e3a0, 0xc00032b620, 0xc00032a890, 0xc0003cc490, 0xc0003f04e0)
	/root/matrixone-main/pkg/sql/compile/compile.go:275 +0x45
created by matrixone/pkg/sql/compile.(*Scope).MergeRun
	/root/matrixone-main/pkg/sql/compile/compile.go:274 +0x1ea

--- a/pkg/hash/setgroup.go
+++ b/pkg/hash/setgroup.go
@@ -393,10 +393,10 @@ func (g *SetGroup) Probe(sels, matched []int64, vecs, gvecs []*vector.Vector, di
 
 func (g *SetGroup) Fill(sels, matched []int64, vecs, gvecs []*vector.Vector,
 	diffs []bool, proc *process.Process) ([]int64, error) {
-	for i, vec := range vecs {
-		switch vec.Typ.Oid {
+	for i, gvec := range gvecs {
+		switch gvec.Typ.Oid {
 		case types.T_int8:
-			gvec := gvecs[i]
+			vec := vecs[i]
 			lnull := vec.Nsp.Any()
 			rnull := gvec.Nsp.Contains(uint64(g.Sel))
 			switch {
@@ -428,7 +428,7 @@ func (g *SetGroup) Fill(sels, matched []int64, vecs, gvecs []*vector.Vector,
 				}
 			}
 		case types.T_int16:
-			gvec := gvecs[i]
+			vec := vecs[i]
 			lnull := vec.Nsp.Any()
 			rnull := gvec.Nsp.Contains(uint64(g.Sel))
 			switch {
@@ -460,7 +460,7 @@ func (g *SetGroup) Fill(sels, matched []int64, vecs, gvecs []*vector.Vector,
 				}
 			}
 		case types.T_int32:
-			gvec := gvecs[i]
+			vec := vecs[i]
 			lnull := vec.Nsp.Any()
 			rnull := gvec.Nsp.Contains(uint64(g.Sel))
 			switch {
@@ -492,7 +492,7 @@ func (g *SetGroup) Fill(sels, matched []int64, vecs, gvecs []*vector.Vector,
 				}
 			}
 		case types.T_int64:
-			gvec := gvecs[i]
+			vec := vecs[i]
 			lnull := vec.Nsp.Any()
 			rnull := gvec.Nsp.Contains(uint64(g.Sel))
 			switch {
@@ -524,7 +524,7 @@ func (g *SetGroup) Fill(sels, matched []int64, vecs, gvecs []*vector.Vector,
 				}
 			}
 		case types.T_uint8:
-			gvec := gvecs[i]
+			vec := vecs[i]
 			lnull := vec.Nsp.Any()
 			rnull := gvec.Nsp.Contains(uint64(g.Sel))
 			switch {
@@ -556,7 +556,7 @@ func (g *SetGroup) Fill(sels, matched []int64, vecs, gvecs []*vector.Vector,
 				}
 			}
 		case types.T_uint16:
-			gvec := gvecs[i]
+			vec := vecs[i]
 			lnull := vec.Nsp.Any()
 			rnull := gvec.Nsp.Contains(uint64(g.Sel))
 			switch {
@@ -588,7 +588,7 @@ func (g *SetGroup) Fill(sels, matched []int64, vecs, gvecs []*vector.Vector,
 				}
 			}
 		case types.T_uint32:
-			gvec := gvecs[i]
+			vec := vecs[i]
 			lnull := vec.Nsp.Any()
 			rnull := gvec.Nsp.Contains(uint64(g.Sel))
 			switch {
@@ -620,7 +620,7 @@ func (g *SetGroup) Fill(sels, matched []int64, vecs, gvecs []*vector.Vector,
 				}
 			}
 		case types.T_uint64:
-			gvec := gvecs[i]
+			vec := vecs[i]
 			lnull := vec.Nsp.Any()
 			rnull := gvec.Nsp.Contains(uint64(g.Sel))
 			switch {
@@ -652,7 +652,7 @@ func (g *SetGroup) Fill(sels, matched []int64, vecs, gvecs []*vector.Vector,
 				}
 			}
 		case types.T_float32:
-			gvec := gvecs[i]
+			vec := vecs[i]
 			lnull := vec.Nsp.Any()
 			rnull := gvec.Nsp.Contains(uint64(g.Sel))
 			switch {
@@ -684,7 +684,7 @@ func (g *SetGroup) Fill(sels, matched []int64, vecs, gvecs []*vector.Vector,
 				}
 			}
 		case types.T_float64:
-			gvec := gvecs[i]
+			vec := vecs[i]
 			lnull := vec.Nsp.Any()
 			rnull := gvec.Nsp.Contains(uint64(g.Sel))
 			switch {
@@ -719,7 +719,7 @@ func (g *SetGroup) Fill(sels, matched []int64, vecs, gvecs []*vector.Vector,
 		case types.T_date:
 		case types.T_datetime:
 		case types.T_char, types.T_json, types.T_varchar:
-			gvec := gvecs[i]
+			vec := vecs[i]
 			lnull := vec.Nsp.Any()
 			rnull := gvec.Nsp.Contains(uint64(g.Sel))
 			switch {

--- a/pkg/sql/build/insert.go
+++ b/pkg/sql/build/insert.go
@@ -56,7 +56,7 @@ func (b *build) buildInsert(stmt *tree.Insert) (op.OP, error) {
 		bat.Vecs[i] = vector.New(typ)
 		delete(mp, attr)
 	}
-	if len(rows.Rows[0]) == 0 {
+	if len(rows.Rows) == 0 || len(rows.Rows[0]) == 0 {
 		return insert.New(id, db, bat, r), nil
 	}
 	for i, vec := range bat.Vecs {

--- a/pkg/sql/build/projection.go
+++ b/pkg/sql/build/projection.go
@@ -9,6 +9,7 @@ import (
 	"matrixone/pkg/sql/op/projection"
 	"matrixone/pkg/sql/tree"
 	"matrixone/pkg/sqlerror"
+	"strings"
 )
 
 func (b *build) checkProjection(ns tree.SelectExprs) error {
@@ -48,6 +49,7 @@ func (b *build) checkProjectionExpr(n tree.Expr, attrs []string) []string {
 		return attrs
 	case *tree.FuncExpr:
 		if name, ok := e.Func.FunctionReference.(*tree.UnresolvedName); ok {
+			name.Parts[0] = strings.ToLower(name.Parts[0])
 			if op, ok := AggFuncs[name.Parts[0]]; ok {
 				if op == aggregation.StarCount {
 					attrs = append(attrs, "*")

--- a/pkg/sql/colexec/aggregation/aggfunc/aggfunc.go
+++ b/pkg/sql/colexec/aggregation/aggfunc/aggfunc.go
@@ -103,6 +103,8 @@ func NewAvg(typ types.Type) aggregation.Aggregation {
 
 func NewSumCount(typ types.Type) aggregation.Aggregation {
 	switch typ.Oid {
+	case types.T_tuple:
+		return sum.NewSumCount(aggregation.ReturnType(aggregation.SumCount, typ))
 	case types.T_float32, types.T_float64:
 		return sum.NewFloatSumCount(aggregation.ReturnType(aggregation.SumCount, typ))
 	case types.T_int8, types.T_int16, types.T_int32, types.T_int64:

--- a/pkg/sql/colexec/aggregation/sum/sumCount.go
+++ b/pkg/sql/colexec/aggregation/sum/sumCount.go
@@ -1,0 +1,54 @@
+package sum
+
+import (
+	"matrixone/pkg/container/types"
+	"matrixone/pkg/container/vector"
+	"matrixone/pkg/sql/colexec/aggregation"
+	"matrixone/pkg/vm/process"
+)
+
+func NewSumCount(typ types.Type) *sumCount {
+	return &sumCount{typ: typ}
+}
+
+func (a *sumCount) Reset() {
+	a.cnt = 0
+	a.sum = 0
+}
+
+func (a *sumCount) Type() types.Type {
+	return a.typ
+}
+
+func (a *sumCount) Dup() aggregation.Aggregation {
+	return &sumCount{typ: a.typ}
+}
+
+func (a *sumCount) Fill(sels []int64, vec *vector.Vector) error {
+	vs := vec.Col.([][]interface{})
+	if n := len(sels); n > 0 {
+		for _, sel := range sels {
+			a.cnt += vs[sel][0].(int64)
+			a.sum += vs[sel][1].(float64)
+		}
+	} else {
+		for _, v := range vs {
+			a.cnt += v[0].(int64)
+			a.sum += v[1].(float64)
+		}
+	}
+	return nil
+}
+
+func (a *sumCount) Eval() interface{} {
+	if a.cnt == 0 {
+		return nil
+	}
+	return []interface{}{a.cnt, a.sum}
+}
+
+func (a *sumCount) EvalCopy(proc *process.Process) (*vector.Vector, error) {
+	vec := vector.New(a.typ)
+	vec.SetCol([][]interface{}{[]interface{}{a.cnt, a.sum}})
+	return vec, nil
+}

--- a/pkg/sql/colexec/aggregation/sum/types.go
+++ b/pkg/sql/colexec/aggregation/sum/types.go
@@ -20,6 +20,12 @@ type floatSum struct {
 	typ types.Type
 }
 
+type sumCount struct {
+	cnt int64
+	sum float64
+	typ types.Type
+}
+
 type intSumCount struct {
 	cnt int64
 	sum int64

--- a/pkg/sql/colexec/mergegroup/group.go
+++ b/pkg/sql/colexec/mergegroup/group.go
@@ -417,6 +417,9 @@ func (ctr *Container) build(n *Argument, proc *process.Process) error {
 				bat.Clean(proc)
 				return err
 			}
+			{
+				fmt.Printf("*********bat: %v\n", bat)
+			}
 			if ctr.bat == nil {
 				ctr.bat = batch.New(true, n.Gs)
 				for i, attr := range n.Gs {

--- a/pkg/sql/compile/compile.go
+++ b/pkg/sql/compile/compile.go
@@ -125,6 +125,9 @@ func (e *Exec) Columns() []*Col {
 func (e *Exec) Run(ts uint64) error {
 	var wg sync.WaitGroup
 
+	fmt.Printf("+++++++++\n")
+	Print(nil, e.ss)
+	fmt.Printf("+++++++++\n")
 	for i := range e.ss {
 		switch e.ss[i].Magic {
 		case Normal:

--- a/pkg/sql/compile/print.go
+++ b/pkg/sql/compile/print.go
@@ -1,0 +1,16 @@
+package compile
+
+import (
+	"fmt"
+	"matrixone/pkg/vm/pipeline"
+)
+
+func Print(prefix []byte, ss []*Scope) {
+	for _, s := range ss {
+		if s.Magic == Merge || s.Magic == Remote {
+			Print(append(prefix, '\t'), s.Ss)
+		}
+		p := pipeline.NewMerge(s.Ins)
+		fmt.Printf("%s:%v %v\n", prefix, s.Magic, p)
+	}
+}

--- a/pkg/sql/handler/handler.go
+++ b/pkg/sql/handler/handler.go
@@ -2,7 +2,6 @@ package handler
 
 import (
 	"bytes"
-	"fmt"
 	"matrixone/pkg/container/batch"
 	"matrixone/pkg/rpcserver/message"
 	"matrixone/pkg/sql/colexec/output"
@@ -48,13 +47,7 @@ func writeBack(u interface{}, bat *batch.Batch) error {
 
 	conn := u.(goetty.IOSession)
 	if bat == nil {
-		{
-			fmt.Printf("write empty\n")
-		}
 		return conn.WriteAndFlush(&message.Message{Sid: 1})
-	}
-	{
-		fmt.Printf("remote bat: %v\n", bat)
 	}
 	if err := protocol.EncodeBatch(bat, &buf); err != nil {
 		return err

